libAvKys/Plugins/Crop/src/cropelement.cpp

```diff
@@ -0,0 +1,258 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Webcamoid is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Webcamoid. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Web-Site: http://webcamoid.github.io/
+ */
+
+#include <cmath>
+#include <QRect>
+#include <QImage>
+#include <QQmlContext>
+#include <akpacket.h>
+#include <akvideopacket.h>
+
+#include "cropelement.h"
+
+class CropElementPrivate
+{
+    public:
+        QRect m_box {}; // satisfies QRect::isNull()
+        QRect m_source {}; // original source size
+        QRect m_maximum {}; // maximum given m_box left,top values
+        QRect m_minimum {}; // smallest acceptable
+        double m_aspectRatio = 0;
+};
+
+CropElement::CropElement(): AkElement()
+{
+    d = new CropElementPrivate;
+}
+
+CropElement::~CropElement()
+{
+    delete d;
+}
+
+QString CropElement::controlInterfaceProvide(const QString &controlId) const
+{
+    Q_UNUSED(controlId);
+    return QString("qrc:/Crop/share/qml/main.qml");
+}
+
+void CropElement::controlInterfaceConfigure(QQmlContext *context, const QString &controlId) const
+{
+    Q_UNUSED(controlId)
+    context->setContextProperty("Cropping", const_cast<QObject *>(qobject_cast<const QObject *>(this)));
+    context->setContextProperty("controlId", this->objectName());
+}
+int CropElement::recalcHeight(const int width)
+{
+    return static_cast<int>(lround(width / d->m_aspectRatio));
+}
+int CropElement::recalcWidth(const int height)
+{
+    return static_cast<int>(lround(height * d->m_aspectRatio));
+}
+
+// getters (using QRectF to enable reading from QML)
+QRectF CropElement::box() const
+{
+    return QRectF(d->m_box);
+}
+QRectF CropElement::minimum() const
+{
+    return QRectF(d->m_minimum);
+}
+QRectF CropElement::maximum() const
+{
+    return QRectF(d->m_maximum);
+}
+double CropElement::aspectRatio() const
+{
+    return d->m_aspectRatio;
+}
+
+// setters (using QRectF to enable setting from QML)
+void CropElement::setMinimum(QRectF qfMin)
+{
+    QRect newMin = qfMin.toRect();
+    if (!newMin.isNull() && newMin.isValid()
+      && d->m_maximum.width() >= newMin.width()
+      && d->m_maximum.height() >= newMin.height()
+      && aspectRatioGood(newMin))
+    {
+        d->m_minimum = newMin;
+        emit minimumChanged(d->m_minimum);
+    }
+}
+
+void CropElement::setBox(QRectF qfBox)
+{
+    bool updateWidth = true;
+    bool updateHeight = true;
+    bool updateMax = false;
+
+    QRect newBox = qfBox.toRect();
+    if (newBox.isNull() || ! newBox.isValid() || newBox == d->m_box)
+        return;
+
+    int maxX = newBox.width() + newBox.left();
```

Comment:
https://github.com/webcamoid/webcamoid/wiki/Coding-style-and-conventions#blocks-of-code-loops-and-conditionals